x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/ExecuteStepsUpdateTask.java

```diff
@@ -88,6 +92,7 @@ public ClusterState execute(ClusterState currentState) throws IOException {
                     if (currentStep.getNextStepKey() == null) {
                         return currentState;
```

Comment:
`nextStepKey` starts out by being null, so this will already do the right thing. I'll add an explicit initialization to `null` in the method though, so it's more apparent.

Reply:
But this task potentially loops through multiple steps right? So if you have a cluster state step , another cluster state step and then `null` next step will get step on the first step and then remain set here when the second step's next step is `null`?

Reply:
Actually I think the behavior might be okay,

- CSAS (cluster state action step) executes with nextStep `A`, moves cluster state to next step (which is also a CSAS)
- CSAS executes with nextStep `null`, returns cluster state at that point
- `clusterStateProcessed` call is invoked, nextStepKey is `A`, but `A` isn't an async action so it's a no-op.

Can you describe the scenario you're worried about in more detail?

Reply:
What you say is correct right now, I am worried that if the logic in `clusterStateProcessed` changes in the future we could inadvertently introduce a bug that will be tricky to track down because the nextStepKey is not actually set to the nextStep at this point. I think we should make it so nextStepKey is correct so we avoid this in the future?

Reply:
Okay, I've re-organized this so it is hopefully clearer